packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,236 @@
+    // downside is having to include an `accountIndex` in the merkle leaf.
+    mapping(uint256 => mapping(address => bool)) public claimed;
+
+    // Index of last seeded root. Next allocation to begin at `lastSeededIndex + 1`.
+    uint256 public lastSeededIndex;
+
+    // Events:
+    event Claimed(address indexed caller, address indexed account, uint256 amount, address indexed rewardToken);
+    event SeededWindow(
+        uint256 indexed windowIndex,
+        uint256 amount,
+        uint256 indexed windowStart,
+        address indexed rewardToken,
+        address owner
+    );
+    event DepositRewards(address indexed owner, uint256 amount);
+    event WithdrawRewards(address indexed owner, uint256 amount);
+    event SetWindowLock(address indexed owner, uint256 indexed windowIndex, bool locked);
+
+    modifier windowNotLocked(uint256 windowIndex) {
+        require(!merkleWindows[windowIndex].locked, "Window distributions locked");
+        _;
+    }
+
+    constructor(address _timerAddress) public Testable(_timerAddress) {}
+
+    /****************************
+     *
+     * Admin functions
+     *
+     ****************************/
+
+    // Set merkle root for the next available window index and seed allocations. Callable by owner of this
+    // contract. Importantly, we assume that the owner of this contract
+    // correctly chooses an amount `totalRewardsDistributed` that is sufficient
+    // to cover all claims within the `merkleRoot`. Otherwise, a race condition
+    // can be created. This situation can occur because we do not segregate reward balances by window,
+    // for code simplicity purposes.
+    //
+    // Example race situation:
+    //     - Window 1 Tree: Owner sets `totalRewardsDistributed=100` and insert proofs that give
+    //                      claimant A 50 tokens and claimant B 51 tokens. The owner has made an error
+    //                      by not setting the `totalRewardsDistributed` correctly to 101).
+    //     - Window 2 Tree: Owner sets `totalRewardsDistributed=1` and insert proofs that give
+    //                      claimant A 1 token. The owner correctly set `totalRewardsDistributed` this time.
+    //     - At this point contract owns 100 + 1 = 101 tokens. Now, imagine the following sequence:
+    //       (1) Claimant A claims 50 tokens for Window 1, contract now has 101 - 50 = 51 tokens.
+    //       (2) Claimant B claims 51 tokens for Window 1, contract now has 51 - 51 = 0 tokens.
+    //       (3) Claimant A tries to claim 1 token for Window 2 but fails because contract has 0 tokens.
+    //     - In summary, the contract owner created a race for step(2) and step(3) in which the first
+    //       claim would succeed and the second claim would fail, even though both claimants would expect
+    //       their claims to suceed.
+    function setWindowMerkleRoot(
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
```

Comment:
Is there a strong reason to keep the start time? Can you think of any scenarios where we'd need to specify some future start time rather than just starting the distribution immediately?

Reply:
I thought we had discussed the desired functionality of "scheduling" distributions in the future, which is the main power of the window start.

Reply:
BTW removing the `require(currentTime >= windowStart)` saves 5000/96000 gas, so its not inconsequential